### PR TITLE
Add TPU USP context parallelism

### PR DESCRIPTION
**Pull request by huytransformer**
# Description

This PR introduces `context_parallel_strategy=usp` ([USP](https://arxiv.org/abs/2405.07719), Ulysses over ring). Follow up to #4687.

Currently does not support load balancing + sequence packing.

# Tests

`python3 -m pytest tests/unit/configs_value_test.py tests/unit/usp_attention_test.py tests/unit/usp_collective_test.py`

Passed. 

`python3 -m pytest tests/unit/attention_test.py -k usp`

3 passed.

## Performance

llama3-8b, v5p (64 chips), CP64, bf16, synthetic data, global batch 1, no load-balancing

Reporting median step time (s)

|  | 64K | 128K | 256K |
|---|---|---|---|
| all-gather (CP64) | 1.31 | 4.03 | 14.33 |
| ring (CP64) | 1.68 | 4.63 | 15.35 |
| usp (16x4) | 1.31 | 4.02 | 13.92 |
| usp (8x8) | 1.44 | 4.05 | 13.65 |
| usp (32x2) | 1.65 | 4.52 | 14.72 |

Example repro command:

```
python3 -m maxtext.trainers.pre_train.train src/maxtext/configs/base.yml model_name=llama3-8b dataset_type=synthetic enable_checkpointing=False steps=20 per_device_batch_size=0.015625 packing=False attention=flash use_tokamax_splash=True use_jax_splash=False allow_split_physical_axes=True max_target_length=65536 context_parallel_strategy=usp ici_context_parallelism=16 ici_context_ulysses_parallelism=4 context_parallel_load_balance=False run_name=usp16x4_65536 base_output_directory=<output dir>
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/configs/types.py [hunk 1]
```diff
@@ -3779,6 +3794,102 @@ def calculate_global_batch_sizes(per_device_batch_size, expansion_factor, num_de
             "TPU Ulysses attention requires num_kv_heads "
             f"({self.num_kv_heads}) to be divisible by context_parallel_size ({context_parallel_size})."
         )
+    if context_parallel_strategy != "usp" and (
```
Comment (gobbleturk) [inline src/maxtext/configs/types.py]: Isn't context_parallel_strategy=ulysses stil supported? E.g. only ulysses, no ring?

File: src/maxtext/configs/types.py [hunk 2]
```diff
@@ -3779,6 +3794,102 @@ def calculate_global_batch_sizes(per_device_batch_size, expansion_factor, num_de
...
+        self.ici_context_ulysses_parallelism != 1 or self.dcn_context_ulysses_parallelism != 1
+    ):
+      raise ValueError(
+          "ici/dcn_context_ulysses_parallelism was specified, but is only supported when "
+          "context_parallel_strategy='usp'."
+      )
```
Comment (gobbleturk) [inline src/maxtext/configs/types.py]: there is a lot of logic here - can you wrap all of this into a function (style suggestion go/small-functions)

File: src/maxtext/configs/types.py [hunk 3]
```diff
@@ -3781,6 +3796,102 @@ def calculate_global_batch_sizes(per_device_batch_size, expansion_factor, num_de
...
+        raise ValueError(
+            "TPU USP attention requires max_target_length "
+            f"({self.max_target_length}) to be divisible by the total context parallelism "
+            f"({usp_ring_size * usp_ulysses_size})."
+        )
+      if self.max_target_length % (usp_ring_size * usp_ring_size) != 0:
+        raise ValueError(
+            "TPU USP attention requires max_target_length to be divisible by ici_context_parallelism squared."
+        )
+      if self.num_query_heads % usp_ulysses_size != 0:
```
[the remark this answers is not in the document]
Comment (gobbleturk) [reply src/maxtext/configs/types.py]: wow go gemini

File: src/maxtext/configs/types.py (hunk 1 above)
Comment (gobbleturk) [reply src/maxtext/configs/types.py]: If a user wanted to run only ulysses without any ring, what would they set? It is not clear to me

File: src/maxtext/kernels/attention/usp_attention.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    raise ValueError("TPU USP attention does not support attention sinks.")
+  if indexer_mask is not None:
+    raise ValueError("TPU USP attention does not support indexer masks.")
+  if bidirectional_mask is not None:
+    raise ValueError("TPU USP attention does not support bidirectional masks.")
+  if record_max_logits:
+    raise NotImplementedError("TPU USP attention does not support record_max_logits yet.")
+
+
+def with_sequence_axes(axis_names: Any, ring_axis: str, ulysses_axis: str, sequence_dim: int) -> Any:
```
Comment (gobbleturk) [inline src/maxtext/kernels/attention/usp_attention.py]: renmae to shard_by_ring_and_ulysses or shard_by_usp or similar?

File: src/maxtext/layers/attention_op.py [hunk 5]
```diff
@@ -1908,6 +2028,27 @@ def wrap_flash_attention(
         attention_output = ulysses_attention.inverse_ulysses_all_to_all(attention_output, context_axis)
         return attention_output, None
 
+      if use_usp:
```
Comment (gobbleturk) [inline src/maxtext/layers/attention_op.py:2031]: I would also wrap everything under the if in a function following style go/small-functions

File: src/maxtext/configs/types.py (hunk 2 above)
Comment (huytransformer) [reply src/maxtext/configs/types.py]: done!

File: src/maxtext/layers/attention_op.py (hunk 5 above)
Comment (huytransformer) [reply src/maxtext/layers/attention_op.py:2031]: done!

File: src/maxtext/kernels/attention/usp_attention.py (hunk 4 above)
Comment (huytransformer) [reply src/maxtext/kernels/attention/usp_attention.py]: renamed to with_usp_sequence_axes!

File: src/maxtext/configs/types.py (hunk 1 above)
Comment (huytransformer) [reply src/maxtext/configs/types.py]: so for every CP we set ici_context_parallelism, only usp additionally sets the ulysses else rejected at init.

File: src/maxtext/configs/types.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (huytransformer) [reply src/maxtext/configs/types.py]: wow go gemini

[the remark this answers is not in the document]
Comment (gobbleturk) [reply src/maxtext/configs/types.py]: you could add in the TP factor as well but generally no one is using TP anyway. This is just a config check, if a user is setting TP and going over head count they will hit a less user friendly jax sharding error

[the remark this answers is not in the document]
Comment (huytransformer) [reply src/maxtext/configs/types.py]: fwiw I think the TP case doesn't reach jax sharding error. validate_head_sharding already checks the local head at init with test 

https://github.com/AI-Hypercomputer/maxtext/blob/bdafbc926eb59931d071118d77cf9e4a01ff26eb/src/maxtext/kernels/attention/ulysses_attention.py#L186-L196

File: src/maxtext/configs/types.py [hunk 6]
```diff
@@ -2861,6 +2874,102 @@ def _validate_use_te_comm_gemm_overlap(self):
...
+        or self.dcn_context_ulysses_parallelism <= 0
+    ):
+      raise ValueError(
+          "TPU USP attention requires explicit positive ici/dcn context parallelism values; "
+          "inferred (-1) sizes are not supported."
+      )
+    if usp_ring_size <= 1:
+      raise ValueError("TPU USP attention requires ici_context_parallelism > 1 for the ring dimension.")
+    if usp_ulysses_size <= 1:
+      raise ValueError("TPU USP attention requires ici_context_ulysses_parallelism > 1 for the Ulysses dimension.")
```
Comment (NuojCheng) [inline src/maxtext/configs/types.py]: what if ici_context_parallelism=-1 or ici_context_ulyses_parallelism=-1

File: src/maxtext/configs/types.py [hunk 7]
```diff
@@ -3644,6 +3753,8 @@ def calculate_global_batch_sizes(per_device_batch_size, expansion_factor, num_de
         )
     if self.context_sharding not in ("context", "expert"):
       raise ValueError(f"Assigned context_sharding f{self.context_sharding} is not supported.")
+    if self.ulysses_context_sharding != "context_ulysses":
+      raise ValueError(f"Assigned ulysses_context_sharding {self.ulysses_context_sharding} is not supported.")
```
Comment (NuojCheng) [inline src/maxtext/configs/types.py:3758]: we will remove this limit in the future using component sharding!!

File: src/maxtext/kernels/attention/ulysses_attention.py [hunk 8]
```diff
@@ -135,13 +135,14 @@ def validate_dkv_sharding(
     axis_names_kv: Any,
     dkv_dim_q: int,
     dkv_dim_kv: int,
+    attention_label: str,
```
Comment (NuojCheng) [inline src/maxtext/kernels/attention/ulysses_attention.py:138]: help me understand when we wanna use ulysses context, should we use
```
ici_context_parallelism=CP
context_parallelism_strategy=ulysses
```
OR
```
ici_ulysses_context_parallelism=CP
```

File: src/maxtext/configs/base.yml [hunk 9]
```diff
@@ -514,30 +514,30 @@ compile_xla_flags: "" # Compiler options e.g. compile_xla_flags="--xla_tpu_num_s
 # Parallelism
 shard_mode: "auto" # can be either auto or explicit
 custom_mesh_and_rule: "" # replace default mesh and logical rule by specifying yml name under config/mesh_and_rule/.
-mesh_axes: ['diloco', 'data', 'stage', 'fsdp', 'fsdp_transpose', 'context', 'context_autoregressive', 'tensor', 'tensor_sequence', 'expert', 'autoregressive']
+mesh_axes: ['diloco', 'data', 'stage', 'fsdp', 'fsdp_transpose', 'context', 'context_ulysses', 'context_autoregressive', 'tensor', 'tensor_sequence', 'expert', 'autoregressive']
```
Comment (NuojCheng) [inline src/maxtext/configs/base.yml]: please update `mesh_axes` in types.py as well

**Review by NuojCheng — APPROVED** [review]

**Review by gobbleturk — APPROVED** [review]
Can we change the ici axis name context_ulysses to context_usp_ulysses or context_ulysses_usp and have some documentation (maybe in types.py near the context_ulysses definition) about how to set

1) ring on its own
2) ulysses on its own
3) usp

File: src/maxtext/kernels/attention/ulysses_attention.py (hunk 8 above)
Comment (huytransformer) [reply src/maxtext/kernels/attention/ulysses_attention.py:138]: So...
Ulysses: We set ici_context_parallelism=CP and context_parallel_strategy="ulysses"
Ring: We set ici_context_parallelism=CP and context_parallel_strategy="ring"
Hybrid USP: We set context_parallel_strategy="usp", use ici_context_parallelism for ring dim size, and use ici_context_ulysses_parallelism for Ulysses dim size

File: src/maxtext/configs/base.yml (hunk 9 above)
Comment (huytransformer) [reply src/maxtext/configs/base.yml]: done

File: src/maxtext/configs/types.py (hunk 6 above)
Comment (huytransformer) [reply src/maxtext/configs/types.py]: we would raise a value error (a few lines up)
